lightning/src/routing/scoring.rs

```diff
@@ -460,25 +454,423 @@ impl<T: Time> Readable for ChannelFailure<T> {
+			.unwrap_or(0)
+	}
+
+	fn decayed_offset_msat(&self, offset_msat: u64) -> u64 {
+		self.now.duration_since(*self.last_updated).as_secs()
+			.checked_div(self.half_life.as_secs())
+			.and_then(|decays| offset_msat.checked_shr(decays as u32))
+			.unwrap_or(0)
+	}
+}
+
+impl<L: DerefMut<Target = u64>, T: Time, U: DerefMut<Target = T>> DirectedChannelLiquidity<L, T, U> {
+	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat`.
+	fn failed_at_channel(&mut self, amount_msat: u64) {
+		if amount_msat < self.max_liquidity_msat() {
+			self.set_max_liquidity_msat(amount_msat);
+		}
+	}
+
+	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat` downstream.
+	fn failed_downstream(&mut self, amount_msat: u64) {
+		if amount_msat > self.min_liquidity_msat() {
+			self.set_min_liquidity_msat(amount_msat);
+		}
+	}
+
+	/// Adjusts the channel liquidity balance bounds when successfully routing `amount_msat`.
+	fn successful(&mut self, amount_msat: u64) {
+		let max_liquidity_msat = self.max_liquidity_msat().checked_sub(amount_msat).unwrap_or(0);
+		self.set_max_liquidity_msat(max_liquidity_msat);
+	}
+
+	/// Adjusts the lower bound of the channel liquidity balance in this direction.
+	fn set_min_liquidity_msat(&mut self, amount_msat: u64) {
+		*self.last_updated = self.now;
+		*self.min_liquidity_offset_msat = amount_msat;
+		*self.max_liquidity_offset_msat = if amount_msat > self.max_liquidity_msat() {
+			0
+		} else {
+			self.decayed_offset_msat(*self.max_liquidity_offset_msat)
+		};
+	}
+
+	/// Adjusts the upper bound of the channel liquidity balance in this direction.
+	fn set_max_liquidity_msat(&mut self, amount_msat: u64) {
+		*self.last_updated = self.now;
+		*self.max_liquidity_offset_msat = self.capacity_msat.checked_sub(amount_msat).unwrap_or(0);
+		*self.min_liquidity_offset_msat = if amount_msat < self.min_liquidity_msat() {
+			0
+		} else {
+			self.decayed_offset_msat(*self.min_liquidity_offset_msat)
+		}
+	}
+}
+
+impl<G: Deref<Target = NetworkGraph>, T: Time> Score for ProbabilisticScorerUsingTime<G, T> {
+	fn channel_penalty_msat(
+		&self, short_channel_id: u64, amount_msat: u64, capacity_msat: u64, source: &NodeId,
+		target: &NodeId
+	) -> u64 {
+		if *source == self.node_id || *target == self.node_id {
+			return 0;
+		}
+
+		let liquidity_offset_half_life = self.params.liquidity_offset_half_life;
+		let success_probability = self.channel_liquidities
+			.get(&short_channel_id)
+			.unwrap_or(&ChannelLiquidity::new())
+			.as_directed(source, target, capacity_msat, liquidity_offset_half_life)
+			.success_probability(amount_msat);
+		match success_probability {
+			Probability::Zero => u64::max_value(),
+			Probability::One => 0,
+			Probability::Ratio { numerator, denominator } => {
+				let success_probability = numerator as f64 / denominator as f64;
+				(-(success_probability.log10()) * amount_msat as f64) as u64
```

Comment:
Right, we have to normalize, but to normalize we'd want to make it a knob, I think - "msats charged per log(failure %)", though. The lack of knob here means users aren't able to control by how much they want to avoid failures, ie they may be willing to take a little longer to pay by retrying a few times in order to avoid paying additional fees. Multiplying by the amount being sent doesn't seem like a useful "penalty per log(failure rate)" answer.

Reply:
Added a `liquidity_penalty_multiplier_msat` parameter and used that instead of the amount.